tachyon/math/finite_fields/mersenne31/packed_mersenne31_neon.cc

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2022 The Plonky3 Authors
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.plonky3 and the LICENCE-APACHE.plonky3
+// file.
+
+#include "tachyon/math/finite_fields/mersenne31/packed_mersenne31_neon.h"
+
+#include <arm_neon.h>
+
+#include "tachyon/math/finite_fields/packed_prime_field32_neon.h"
+
+namespace tachyon::math {
+
+namespace {
+
+uint32x4_t kP;
+uint32x4_t kZero;
+uint32x4_t kOne;
+
+uint32x4_t ToVector(const PackedMersenne31Neon& packed) {
+  return vld1q_u32(reinterpret_cast<const uint32_t*>(packed.values().data()));
+}
+
+PackedMersenne31Neon FromVector(uint32x4_t vector) {
+  PackedMersenne31Neon ret;
+  vst1q_u32(reinterpret_cast<uint32_t*>(ret.values().data()), vector);
+  return ret;
+}
+
+uint32x4_t Add(uint32x4_t lhs, uint32x4_t rhs) {
+  return AddMod32(lhs, rhs, kP);
+}
+
+uint32x4_t Sub(uint32x4_t lhs, uint32x4_t rhs) {
+  return SubMod32(lhs, rhs, kP);
+}
+
+uint32x4_t Neg(uint32x4_t val) { return NegMod32(val, kP); }
+
+// Multiply two 31-bit numbers to obtain a 62-bit immediate result, and return
+// the high 31 bits of that result. Results are arbitrary if the inputs do not
+// fit in 31 bits.
+uint32x4_t mul_31x31_to_hi_31(uint32x4_t lhs, uint32x4_t rhs) {
+  return vreinterpretq_u32_s32(
+      vqdmulhq_s32(vreinterpretq_s32_u32(lhs), vreinterpretq_s32_u32(rhs)));
+}
+
+uint32x4_t Mul(uint32x4_t lhs, uint32x4_t rhs) {
+  // We want this to compile to:
+  //      sqdmulh  prod_hi31.4s, lhs.4s, rhs.4s
+  //      mul      t.4s, lhs.4s, rhs.4s
+  //      mla      t.4s, prod_hi31.4s, P.4s
+  //      sub      u.4s, t.4s, P.4s
+  //      umin     r.4s, t.4s, u.4s
+  // throughput: 1.25 cyc/vec (3.2 els/cyc)
+  // latency: 10 cyc
+
+  // Let prod := lhs * rhs
+  //     prod_hi31 := (prod >> 2³¹) (mod 2³¹)
+  //               = a₃₁ + 2 * a₃₂ + 2² * a₃₃ + ... + 2³⁰ * a₆₂
```

Comment:
I believe the superscript unicode of zero in 2^30 went wrong. It appears just like a tiny box for me. Please check!